app/src/main/kotlin/pl/gov/mc/protego/bluetooth/advertiser/ProteGoGattServer.kt

```diff
@@ -0,0 +1,324 @@
+
+        val gattCharacteristic = BluetoothGattCharacteristic(
+            UUID.fromString(ProteGoCharacteristicUUIDString),
+            BluetoothGattCharacteristic.PROPERTY_READ or BluetoothGattCharacteristic.PROPERTY_WRITE,
+            BluetoothGattCharacteristic.PERMISSION_READ or BluetoothGattCharacteristic.PERMISSION_WRITE
+        )
+
+        val gattService = BluetoothGattService(
+            UUID.fromString(ProteGoServiceUUIDString),
+            BluetoothGattService.SERVICE_TYPE_PRIMARY
+        )
+
+        if (!gattService.addCharacteristic(gattCharacteristic)) {
+            Timber.d("Failed to add characteristic")
+            gattServer.close()
+            return ServerResult.Failure.CannotAddCharacteristic
+        }
+
+        if (!gattServer.addService(gattService)) {
+            Timber.d("Failed to add service")
+            gattServer.close()
+            return ServerResult.Failure.CannotAddService
+        }
+
+        return null
+    }
+
+    fun close(): Boolean {
+        val gattServer = this.gattServer
+        if (gattServer == null) {
+            Timber.d("GATT server already closed")
+            return false
+        }
+
+        pendingWrites.clear()
+        rssiLatches.values.forEach { it.cancel() }
+        rssiLatches.clear()
+        gattServer.close()
+        this.gattServer = null
+        Timber.d("GATT server closed")
+        return true
+    }
+
+    // GATT Server callbacks -----------------------------------------------------------------------
+
+    override fun onDescriptorReadRequest(
+        device: BluetoothDevice?,
+        requestId: Int,
+        offset: Int,
+        descriptor: BluetoothGattDescriptor?
+    ) {
+        super.onDescriptorReadRequest(device, requestId, offset, descriptor)
+        Timber.d("onDescriptorReadRequest, device=${device?.address}, requestId=${requestId}, offset=${offset}, desc=${descriptor?.uuid}")
+        val gattServer = this.gattServer
+        if (gattServer == null || device == null) {
+            Timber.e("invalid onDescriptorReadRequest params")
+            return
+        }
+        gattServer.sendResponse(device, requestId, BluetoothGatt.GATT_READ_NOT_PERMITTED, offset, null)
+    }
+
+    override fun onNotificationSent(device: BluetoothDevice?, status: Int) {
+        super.onNotificationSent(device, status)
+        Timber.d("onNotificationSent, device=${device?.address}, status=${status}")
+        val gattServer = this.gattServer
+        if (gattServer == null || device == null) {
+            Timber.e("invalid onNotificationSent params")
+            return
+        }
+    }
+
+    override fun onMtuChanged(device: BluetoothDevice?, mtu: Int) {
+        super.onMtuChanged(device, mtu)
+        Timber.d("onMtuChanged, device=${device?.address}, mtu=${mtu}")
+        val gattServer = this.gattServer
+        if (gattServer == null || device == null) {
+            Timber.e("invalid onMtuChanged params")
```

Comment:
Czemu brak serwera oznacza niepoprawne parametry? Właściwie tyczy się to wszystkich metod.

Reply:
Wydaje mi się, że chodzi o to, że nie powinno być takiego przypadku.

Reply:
Domyślam się, ale log sugeruje coś zupełnie innego.

Reply:
Potrzebuję sugestii w takim razie

Reply:
`unable to process onMtuChanged`? `error encountered in onMtuChanged`? `invalid state in onMtuChanged`?
Chodzi mi tylko o to, że gattServer stanowczo nie jest parametrem, więc ten log jest mylący. Może rozbić warunki na rozdzielne i oddzielnie zalogować jeśli gattServer jest nullem?